### PR TITLE
Profiling of neural networks

### PR DESCRIPTION
**Pull request by bedapisl**
Profiling is switch on by `model.profile` variable in YAML config:
```
model:
    profile: True
```
Profile is saved in the `log_dir` of the model and can be viewed in Google Chrome at address `chrome://tracing/`.

### CODE REVIEW

**Comment by blazekadam** [comment]
Whoa, @FloopCZ I thought we already gave up on `"` vs `'` thing...

File: emloop_tensorflow/frozen_model.py [hunk 1]
```diff
@@ -83,7 +91,19 @@ def run(self, batch: el.Batch, train: bool=False, stream: el.datasets.StreamWrap
...
+        if self._profile:
+            run_options = tf.RunOptions(trace_level=tf.RunOptions.FULL_TRACE)
+            run_metadata = tf.RunMetadata()
+
+            outputs = self._session.run(fetches=fetches, feed_dict=feed_dict,
+                                        options=run_options, run_metadata=run_metadata)
+
+            with open(path.join(self._log_dir, "profile.json"), "w") as ofile:
+                tl = timeline.Timeline(run_metadata.step_stats)
+                ofile.write(tl.generate_chrome_trace_format())
```
[the remark this answers is not in the document]
Comment (blazekadam) [reply emloop_tensorflow/frozen_model.py]: Well, this is a bit tricky. Both first (warm-up) and last (possibly smaller batch) profiles may be inaccurate. So what do we want to actually save? I guess keeping last `keep_profiles: int = 5` is reasonable.

File: emloop_tensorflow/model.py [hunk 2]
```diff
@@ -44,7 +45,7 @@ def __init__(self,  # pylint: disable=too-many-arguments
                  dataset: Optional[el.AbstractDataset], log_dir: Optional[str], inputs: List[str], outputs: List[str],
```
Comment (blazekadam) [inline emloop_tensorflow/model.py]: I would make the `log_dir` indeed optional with default `None`.

File: emloop_tensorflow/model.py [hunk 3]
```diff
@@ -82,6 +83,7 @@ def __init__(self,  # pylint: disable=too-many-arguments
         :param monitor: monitor signal mean and variance of the tensors which names contain the specified value
         :param restore_fallback: ignored arg. (allows training from configs saved by emloop where it is added)
         :param clip_gradient: limit the absolute value of the gradient; set to None for no clipping
+        :param profile: whether profile.json should be saved to log_dir
```
Comment (blazekadam) [inline emloop_tensorflow/model.py]: ditto description

File: emloop_tensorflow/model.py [hunk 4]
```diff
@@ -223,12 +226,25 @@ def run(self, batch: el.Batch, train: bool=False, stream: el.datasets.StreamWrap
                 for output_name in self.output_names:
                     fetches.append(tower[output_name])
 
+        run_options = None
```
Comment (blazekadam) [inline emloop_tensorflow/model.py]: This is more or less the same code as in `FrozenModel`. Can you perhaps wrap `session.run` to some util function which would be utilized in both classes?

File: emloop_tensorflow/model.py (hunk 2 above)
Comment (bedapisl) [reply emloop_tensorflow/model.py]: Is this change you propose in some way related to this pull request?

File: docs/conf.py [hunk 5]
```diff
@@ -13,7 +13,7 @@
 # General information about the project.                         
 project = 'emloop-tensorflow' 
 copyright = '2018, Iterait a.s.'                     
-author = 'Blazek Adam, Belohlavek Petr, Matzner Filip'
+author = 'Blazek Adam, Belohlavek Petr, Matzner Filip, Bedrich Pisl'
```
Comment (blazekadam) [inline docs/conf.py]: nice :)

File: docs/advanced.rst [hunk 6]
```diff
@@ -0,0 +1,27 @@
+Profiling networks
```
Comment (blazekadam) [inline docs/advanced.rst]: Thank you for the docs.

File: docs/conf.py (hunk 5 above)
Comment (FloopCZ) [reply docs/conf.py]: Nice except that Beda's name and surname are in the wrong order. :-)

Comment (blazekadam) [reply docs/conf.py]: whatever

**Pull request merged**